packages/nextjs/src/client/index.tsx

```diff
@@ -16,22 +18,27 @@ type NextClerkProviderProps = {
   Partial<PublishableKeyOrFrontendApi>;
 
 export function ClerkProvider({ children, ...rest }: NextClerkProviderProps): JSX.Element {
-  // @ts-expect-error
   // Allow for overrides without making the type public
-  const { authServerSideProps, frontendApi, publishableKey, proxyUrl, __clerk_ssr_state, clerkJSUrl, ...restProps } =
-    rest;
+  // @ts-expect-error
+  // prettier-ignore
+  const { authServerSideProps, frontendApi, publishableKey, proxyUrl, __clerk_ssr_state, clerkJSUrl, navigate: userNavigate, ...restProps } = rest;
   const { push } = useRouter();
 
   ReactClerkProvider.displayName = 'ReactClerkProvider';
 
+  const navigate = async (to: string) => {
+    await invalidateNextRouterNon200ResponseCache();
+    return (userNavigate || push)(to);
+  };
+
   return (
     <ReactClerkProvider
       frontendApi={frontendApi || process.env.NEXT_PUBLIC_CLERK_FRONTEND_API || ''}
       publishableKey={publishableKey || process.env.NEXT_PUBLIC_CLERK_PUBLISHABLE_KEY || ''}
       clerkJSUrl={clerkJSUrl || process.env.NEXT_PUBLIC_CLERK_JS}
       // @ts-expect-error
       proxyUrl={proxyUrl || process.env.NEXT_PUBLIC_CLERK_PROXY_URL}
-      navigate={to => push(to)}
+      navigate={navigate}
```

Comment:
I feel like we do need before/after setActive hooks, otherwise we risk ending up on the same flow if a setActive is called with a beforeEmit and the middleware somehow triggers afterwards

Reply:
If we're unsure about the api (I think that's the case) and since I don' want to add props accepting non-scaler values to ClerkProvider (/app requires the values to be serializable), Im thinking we can try the following API:

https://github.com/clerkinc/javascript/blob/2c0112908e2be131bd951ad875e4e48621779162/packages/clerk-js/src/core/fapiClient.ts#L74

cc @colinclerk

Reply:
Update: We're going to completely nuke the cache and move the logic in the beforeSetActive handler